TeleIRC crashing

I get these errors a lot.  Usually I can just run teleirc again and it starts fine. But after an hour or so it crashes with the same error.

$ ./teleirc
INFO: 2025/02/05 00:31:18 Current TeleIRC version: v2.3.0-10-g0af3dfe
INFO: 2025/02/05 00:31:18 Creating new Telegram bot client...
INFO: 2025/02/05 00:31:18 Creating new IRC bot client...
INFO: 2025/02/05 00:31:18 Starting up Telegram bot...
INFO: 2025/02/05 00:31:18 Starting up IRC bot...
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x10e839f]

goroutine 23 [running]:
github.com/go-telegram-bot-api/telegram-bot-api.(*BotAPI).MakeRequest(0x0, {0x11e77c1, 0xb}, 0xc0002c7a70)
        C:/Users/biger/go/pkg/mod/github.com/go-telegram-bot-api/telegram-bot-api@v4.6.4+incompatible/bot.go:64 +0x7f
github.com/go-telegram-bot-api/telegram-bot-api.(*BotAPI).makeMessageRequest(_, {_, _}, _)
        C:/Users/biger/go/pkg/mod/github.com/go-telegram-bot-api/telegram-bot-api@v4.6.4+incompatible/bot.go:119 +0x7b
github.com/go-telegram-bot-api/telegram-bot-api.(*BotAPI).sendChattable(_, {_, _})
        C:/Users/biger/go/pkg/mod/github.com/go-telegram-bot-api/telegram-bot-api@v4.6.4+incompatible/bot.go:343 +0xd0
github.com/go-telegram-bot-api/telegram-bot-api.(*BotAPI).Send(_, {_, _})
        C:/Users/biger/go/pkg/mod/github.com/go-telegram-bot-api/telegram-bot-api@v4.6.4+incompatible/bot.go:274 +0xb8
github.com/ritlug/teleirc/internal/handlers/telegram.(*Client).SendMessage(0xc0001cce80, {0xc0002410c0?, 0xc000305c50?})
        C:/dev/src/teleirc/internal/handlers/telegram/telegram.go:37 +0xc8
github.com/ritlug/teleirc/internal/handlers/irc.Client.SendToTg(...)
        C:/dev/src/teleirc/internal/handlers/irc/irc.go:123
github.com/ritlug/teleirc/internal/handlers/irc.disconnectHandler.func1(0xf232cf?, {0x0, 0x0, {0x0, 0x0, 0x0}, {0x11eae0f, 0x13}, {0xc000289260, 0x1, ...}, ...})
        C:/dev/src/teleirc/internal/handlers/irc/handlers.go:105 +0x124
github.com/lrstanley/girc.HandlerFunc.Execute(0xe665e9?, 0xc000305e00?, {0x0, 0x0, {0x0, 0x0, 0x0}, {0x11eae0f, 0x13}, {0xc000289260, ...}, ...})
        C:/Users/biger/go/pkg/mod/github.com/lrstanley/girc@v0.0.0-20190801035559-4fc93959e1a7/handler.go:67 +0x43
github.com/lrstanley/girc.(*Caller).exec.func1(0x0)
        C:/Users/biger/go/pkg/mod/github.com/lrstanley/girc@v0.0.0-20190801035559-4fc93959e1a7/handler.go:231 +0x3d8
created by github.com/lrstanley/girc.(*Caller).exec in goroutine 8
        C:/Users/biger/go/pkg/mod/github.com/lrstanley/girc@v0.0.0-20190801035559-4fc93959e1a7/handler.go:209 +0x61b

@Tjzabel Well I just ran teleirc with the debug switch and got the error immediately:

biger@aceraspireone MINGW64 /c/dev/src/teleirc (main)
$ ./teleirc -debug=true
INFO: 2025/02/05 21:30:04 Current TeleIRC version: v2.3.0-10-g0af3dfe
DEBUG: 2025/02/05 21:30:04 Debug mode enabled!
INFO: 2025/02/05 21:30:04 Creating new Telegram bot client...
INFO: 2025/02/05 21:30:04 Creating new IRC bot client...
INFO: 2025/02/05 21:30:04 Starting up Telegram bot...
INFO: 2025/02/05 21:30:04 Starting up IRC bot...
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: JOIN
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: NICK
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: TOPIC
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: CLIENT_DISCONNECTED
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: KICK
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: PRIVMSG
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: PART
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: QUIT
DEBUG: 2025/02/05 21:30:04 Adding IRC event handler: CLIENT_CONNECTED
INFO: 2025/02/05 21:30:05 Authorized on account TeleIRC2Bot
ERROR: 2025/02/05 21:30:05 Missing message data
DEBUG: 2025/02/05 21:30:05 messageHandler triggered
DEBUG: 2025/02/05 21:30:05 messageHandler triggered
DEBUG: 2025/02/05 21:30:05 messageHandler triggered
DEBUG: 2025/02/05 21:30:05 messageHandler triggered
DEBUG: 2025/02/05 21:30:07 messageHandler triggered
DEBUG: 2025/02/05 21:30:09 messageHandler triggered
DEBUG: 2025/02/05 21:30:11 messageHandler triggered
DEBUG: 2025/02/05 21:30:13 messageHandler triggered
ERROR: 2025/02/05 21:30:21 Missing message data
DEBUG: 2025/02/05 21:30:21 photoHandler triggered
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x80 pc=0xdab512]

goroutine 7 [running]:
github.com/ritlug/teleirc/internal/handlers/telegram.getImgurLink(0xc0002b10c0, {0xc00000e2a0, 0x63})
        C:/dev/src/teleirc/internal/handlers/telegram/imgur.go:101 +0x8d2
github.com/ritlug/teleirc/internal/handlers/telegram.uploadImage(0xc0002b10c0, {0x15f7b23, 0xc00047cc80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...})
        C:/dev/src/teleirc/internal/handlers/telegram/helpers.go:68 +0x137
github.com/ritlug/teleirc/internal/handlers/telegram.photoHandler(0xc0002b10c0, {0x15f7b23, 0xc00047cc80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...})
        C:/dev/src/teleirc/internal/handlers/telegram/handler.go:162 +0x58
github.com/ritlug/teleirc/internal/handlers/telegram.updateHandler(0xc0002b10c0, 0xc0000caa10)
        C:/dev/src/teleirc/internal/handlers/telegram/handler.go:45 +0x2ca
github.com/ritlug/teleirc/internal/handlers/telegram.(*Client).StartBot(0xc0002b10c0, 0xc00000e620, 0xc0002b1100)
        C:/dev/src/teleirc/internal/handlers/telegram/telegram.go:78 +0x387
created by main.main in goroutine 1
        C:/dev/src/teleirc/cmd/teleirc.go:53 +0x666

biger@aceraspireone MINGW64 /c/dev/src/teleirc (main)
$

Usually what happens at this point, is I just run it again and it works without crashing. For an hour or so, then it crashes.

I determined that the crash is happening when a message containing a picture is sent in the Telegram group. It crashes on the getImgurLink method. I tried creating my own API token according to the FAQ page- no difference.

DEBUG: 2025/02/07 00:00:18 photoHandler triggered
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x80 pc=0xe6b512]

goroutine 20 [running]:
github.com/ritlug/teleirc/internal/handlers/telegram.getImgurLink(0xc00031e540, {0xc0000cdf10, 0x63})
        C:/dev/src/teleirc/internal/handlers/telegram/imgur.go:101 +0x8d2
github.com/ritlug/teleirc/internal/handlers/telegram.uploadImage(0xc00031e540, {0x15f7c22, 0xc000404140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...})
        C:/dev/src/teleirc/internal/handlers/telegram/helpers.go:68 +0x137
github.com/ritlug/teleirc/internal/handlers/telegram.photoHandler(0xc00031e540, {0x15f7c22, 0xc000404140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...})
        C:/dev/src/teleirc/internal/handlers/telegram/handler.go:162 +0x58
github.com/ritlug/teleirc/internal/handlers/telegram.updateHandler(0xc00031e540, 0xc00019b1f0)
        C:/dev/src/teleirc/internal/handlers/telegram/handler.go:45 +0x2ca
github.com/ritlug/teleirc/internal/handlers/telegram.(*Client).StartBot(0xc00031e540, 0xc0000cc2a0, 0xc00031e580)
        C:/dev/src/teleirc/internal/handlers/telegram/telegram.go:78 +0x387
created by main.main in goroutine 1
        C:/dev/src/teleirc/cmd/teleirc.go:53 +0x666

--- a/internal/handlers/telegram/handler.go
+++ b/internal/handlers/telegram/handler.go
@@ -165,9 +165,14 @@ func photoHandler(tg *Client, u tgbotapi.Update) {
 	if caption == "" {
 		caption = "No caption provided."
 	}
-	formatted := "'" + caption + "' uploaded by " + username + ": " + link
 
-	tg.sendToIrc(formatted)
+	// TeleIRC can fail to upload to Imgur
+	if link == "" {
+		tg.logger.LogError("Failed imgur photo upload for", username)
+	} else {
+		formatted := "'" + caption + "' uploaded by " + username + ": " + link
+		tg.sendToIrc(formatted)
+	}
 }
 
 /*

--- a/internal/handlers/telegram/imgur.go
+++ b/internal/handlers/telegram/imgur.go
@@ -98,6 +98,11 @@ func getImgurLink(tg *Client, tgLink string) string {
 		tg.logger.LogError("Could not retrieve imgur json data:", err)
 	}
 
+    if resp.Data == nil {
+        tg.logger.LogError("Imgur response data was null. Possible API rate limiting.")
+        return ""
+    }
+
 	if resp.Data.Deletehash != "" {
 		deleteLink := "https://imgur.com/delete/" + resp.Data.Deletehash
 		tg.logger.LogInfo("Deletion link for", resp.Data.Link, "is", deleteLink)

--- a/internal/handlers/telegram/imgur.go
+++ b/internal/handlers/telegram/imgur.go
@@ -98,10 +98,10 @@ func getImgurLink(tg *Client, tgLink string) string {
 		tg.logger.LogError("Could not retrieve imgur json data:", err)
 	}
 
-    if resp.Data == nil {
-        tg.logger.LogError("Imgur response data was null. Possible API rate limiting.")
-        return ""
-    }
+	if resp.Data == nil {
+		tg.logger.LogError("Imgur response data was null. Possible API rate limiting.")
+		return ""
+	}
 
 	if resp.Data.Deletehash != "" {
 		deleteLink := "https://imgur.com/delete/" + resp.Data.Deletehash

